operator/internal/handlers/internal/rules/secrets.go

```diff
@@ -0,0 +1,37 @@
+package rules
+
+import (
+	"github.com/ViaQ/logerr/v2/kverrors"
+
+	lokiv1beta1 "github.com/grafana/loki/operator/api/v1beta1"
+	"github.com/grafana/loki/operator/internal/manifests"
+
+	corev1 "k8s.io/api/core/v1"
+)
+
+// ExtractRulerSecret reads a k8s secret infto a ruler secret struct if valid.
+func ExtractRulerSecret(s *corev1.Secret, t lokiv1beta1.RemoteWriteAuthType) (*manifests.RulerSecret, error) {
+	switch t {
+	case lokiv1beta1.BasicAuthorization:
+		username, ok := s.Data["username"]
+		if !ok {
```

Comment:
Should we check for empty strings here? (same for password & token)
```suggestion
		username := s.Data["username"]
		if len(username) == 0 {
```